### ISSUE TITLE
2节点训练13B LLaMA模型效率只能达到840 token/sec/GPU

### ISSUE THREAD

**Issue by YaboSun** [alibaba/Megatron-LLaMA#16]
基于2台A800x80G训练13B LLaMA模型发现效率只能达到840 token/sec/GPU，不知道是什么原因，详细配置如下：
    --tensor-model-parallel-size 4 \
    --pipeline-model-parallel-size 1 \ 
    --sequence-parallel \
    --distributed-timeout-minutes 60 \
    --no-position-embedding \
    --use-rotary-position-embeddings \
    --RMSNorm \
    --layernorm-epsilon 1e-6 \
    --causal-lm \
    --disable-bias-linear \
    --swiglu \
    --micro-batch-size 4 \
    --global-batch-size 1024 \
    --train-iters 50862 \
    --lr 6.0e-5 \
    --lr-decay-iters 32000 \
    --lr-decay-style cosine \
    --min-lr 6e-6 \
    --weight-decay 0.1 \
    --lr-warmup-fraction .01 \
    --clip-grad 1.0 \
    --override-opt_param-scheduler \
    --overlapped-distributed-optimizer \
    --reduce-bucket-size=4e8 \
    --no-gradient-accumulation-fusion \
    --dataloader-type cyclic \

**Comment by li-yi-dong** [comment]
两台的意思是两台8xA100？

**Comment by YaboSun** [comment]
> 两台的意思是两台8xA100？

两台8×A800，不是A100

**Comment by li-yi-dong** [comment]
可以把完整的配置发出来吗

**Comment by YaboSun** [comment]
> 可以把完整的配置发出来吗
    --num-layers 40 \
    --hidden-size 5120 \
    --num-attention-heads 40 \
    --seq-length 2048 \
    --max-position-embeddings 2048
    --tensor-model-parallel-size 4 \
    --pipeline-model-parallel-size 1 \
    --sequence-parallel \
    --distributed-timeout-minutes 60 \
    --no-position-embedding \
    --use-rotary-position-embeddings \
    --RMSNorm \
    --layernorm-epsilon 1e-6 \
    --causal-lm \
    --disable-bias-linear \
    --swiglu \
    --micro-batch-size 4 \
    --global-batch-size 1024 \
    --train-iters 50862 \
    --lr 6.0e-5 \
    --lr-decay-iters 32000 \
    --lr-decay-style cosine \
    --min-lr 6e-6 \
    --weight-decay 0.1 \
    --lr-warmup-fraction .01 \
    --clip-grad 1.0 \
    --override-opt_param-scheduler \
    --overlapped-distributed-optimizer \
    --reduce-bucket-size=4e8 \
    --no-gradient-accumulation-fusion \
    --dataloader-type cyclic \
    --tokenizer-type PretrainedFromHF \
    --tokenizer-name-or-path ${TOKENIZER_MODEL} \
    --data-impl mmap \
    --split 998,2,1 \
    --log-interval 1 \
    --timing-log-option=all \
    --timing-log-level=2 \
    --save-interval 100000 \
    --eval-interval 100 \
    --eval-iters 10 \
    --load ${LOAD_DIR} \
    --recompute-activations \
    --use-distributed-optimizer \
    --recompute-granularity selective \
    --bf16

**Comment by YaboSun** [comment]
> 可以把完整的配置发出来吗

--num-layers 40
--hidden-size 5120
--num-attention-heads 40
--seq-length 2048
--max-position-embeddings 2048
--tensor-model-parallel-size 4
--pipeline-model-parallel-size 1
--sequence-parallel
--distributed-timeout-minutes 60
--no-position-embedding
--use-rotary-position-embeddings
--RMSNorm
--layernorm-epsilon 1e-6
--causal-lm
--disable-bias-linear
--swiglu
--micro-batch-size 4
--global-batch-size 1024
--train-iters 50862
--lr 6.0e-5
--lr-decay-iters 32000
--lr-decay-style cosine
--min-lr 6e-6
--weight-decay 0.1
--lr-warmup-fraction .01
--clip-grad 1.0
--override-opt_param-scheduler
--overlapped-distributed-optimizer
--reduce-bucket-size=4e8
--no-gradient-accumulation-fusion
--dataloader-type cyclic
--tokenizer-type PretrainedFromHF
--tokenizer-name-or-path ${TOKENIZER_MODEL}
--data-impl mmap
--split 998,2,1
--log-interval 1
--timing-log-option=all
--timing-log-level=2
--save-interval 100000
--eval-interval 100
--eval-iters 10
--load ${LOAD_DIR}
--recompute-activations
--use-distributed-optimizer
--recompute-granularity selective
--bf16

**Comment by li-yi-dong** [comment]
没用flash-attention，而且--use-distributed-optimizer 和--overlapped-distributed-optimizer 不能同时用

**Comment by YaboSun** [comment]
> 没用flash-attention，而且--use-distributed-optimizer 和--overlapped-distributed-optimizer 不能同时用

添加了flash-attention，并且只开启--overlapped-distributed-optimizer最终是1162 tokens/sec/gpu，比预期还是低很多，详细配置如下：

```
--num-layers 40
--hidden-size 5120
--num-attention-heads 40
--seq-length 2048
--max-position-embeddings 2048
--tensor-model-parallel-size 4
--pipeline-model-parallel-size 1
--sequence-parallel
--distributed-timeout-minutes 60
--no-position-embedding
--use-rotary-position-embeddings
--RMSNorm
--layernorm-epsilon 1e-6
--causal-lm
--disable-bias-linear
--swiglu
--micro-batch-size 4
--global-batch-size 1024
--train-iters 50862
--lr 6.0e-5
--lr-decay-iters 32000
--lr-decay-style cosine
--min-lr 6e-6
--weight-decay 0.1
--lr-warmup-fraction .01
--clip-grad 1.0
--override-opt_param-scheduler
--reduce-bucket-size=4e8
--no-gradient-accumulation-fusion
--dataloader-type cyclic
--recompute-activations 
--overlapped-distributed-optimizer 
--use-flash-attn 
--bf16
--data-path $DATA_PATH 
--tokenizer-type PretrainedFromHF 
--tokenizer-name-or-path ${TOKENIZER_MODEL} 
--data-impl mmap 
--split 998,2,1
--log-interval 1 
--timing-log-option=all 
--timing-log-level=2 
--save-interval 100000 
--eval-interval 100 
--eval-iters 10 
--load ${LOAD_DIR}
```

**Comment by li-yi-dong** [comment]
两台机器间的带宽是多少

**Comment by YaboSun** [comment]
4张IB网卡通信，单张200G/s

**Comment by li-yi-dong** [comment]
换成这个呢
--use-distributed-optimizer

**Comment by YaboSun** [comment]
> 换成这个呢 --use-distributed-optimizer

换成这个实测是1621

**Comment by li-yi-dong** [comment]
梯度聚合较大的情况下--use-distributed-optimizer 会快一些，不过您--use-distributed-optimizer 的速度也偏慢。
想进一步看看的话需要profile 一下。

**Comment by harveylihr** [comment]
同样的问题，2机16卡速度比1机8卡速度还慢，同样的global batch size。

**Comment by li-yi-dong** [comment]
> 同样的问题，2机16卡速度比1机8卡速度还慢，同样的global batch size。

可以发一下启动的脚本吗，如果可以的话最好能用torch 的profiler trace 一下看看
